{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a5dd7d4-467d-4aba-9355-04b482638ea7",
   "metadata": {},
   "source": [
    "# Scop3P\n",
    "\n",
    "A comprehensive database of human phosphosites within their full context. Scop3P integrates sequences (UniProtKB/Swiss-Prot), structures (PDB), and uniformly reprocessed phosphoproteomics data (PRIDE) to annotate all known human phosphosites. \n",
    "\n",
    "Scop3P, available at https://iomics.ugent.be/scop3p, presents a unique resource for visualization and analysis of phosphosites and for understanding of phosphosite structure–function relationships.\n",
    "\n",
    "Please cite: https://doi.org/10.1021/acs.jproteome.0c00306"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb1e9ff-c789-4c96-b71d-5da2af53b437",
   "metadata": {},
   "source": [
    "## Install Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba367321-b7ac-44b9-98c0-c800fd1ae008",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!jupyter labextension install jupyterlab_3dmol\n",
    "!pip install pandas matplotlib py3Dmol b2btools==3.0.7b2 pyvis ipycytoscape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "62a30299-1084-4700-a8ed-64f77b68c0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "import requests, tempfile,json\n",
    "import pandas as pd \n",
    "from b2bTools import SingleSeq, constants\n",
    "import py3Dmol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2dc9e1f-97d2-45a8-a21a-b7258664d265",
   "metadata": {},
   "source": [
    "## Target protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bd5e0bc6-c9ee-485f-9a7c-3adc53f7cd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_PROTEIN_ID = \"P07949\" # Write here the Protein ID of your protein of interest\n",
    "PDB_ID = \"2IVT\" # Write here the PDB ID of your protein of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e622d9b-1922-4851-ab6b-41ae0d62c395",
   "metadata": {},
   "source": [
    "## API Request\n",
    "\n",
    "This function makes a GET request to Scop3P API endpoint for a given protein accession ID and returns the protein sequence in string format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87d604d4-ba1a-4acb-94b7-0481abe40fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_protein_modifications(accession):\n",
    "    \"\"\"\n",
    "    Fetches protein modifications for a given UniProt ID.\n",
    "\n",
    "    Parameters:\n",
    "    accession (str): UniProt ID of the protein.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary containing protein modifications.\n",
    "    \"\"\"\n",
    "    BASE_URL = \"https://iomics.ugent.be/scop3p/api/modifications\"\n",
    "    url = f'{BASE_URL}?accession={accession}'\n",
    "    headers = {'accept': 'application/json'}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if response.status_code == 200:\n",
    "        return response.json()\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09ac2da-0a01-4b83-97ba-96c9fa9fb598",
   "metadata": {},
   "source": [
    "## Data parsing\n",
    "\n",
    "For parsing the JSON as a table, we'll use pandas library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fdd2c83a-bbdf-4ccd-993f-92078980fd9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Scop3P: A Comprehensive Resource of Human Phosphosites within Their Full Context\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "RET_HUMAN:Proto-oncogene tyrosine-protein kinase receptor Ret (EC 2.7.10.1) (Cadherin family member 12) (Proto-oncogene c-Ret) [Cleaved into: Soluble RET kinase fragment; Extracellular cell-membrane anchored RET cadherin 120 kDa fragment]\n",
      "\n",
      "Phospho-sites found: 9 entries.\n",
      "\n",
      "Full entry available on SCOP3P website: https://iomics.ugent.be/scop3p/index?protein=P07949\n",
      "\n"
     ]
    }
   ],
   "source": [
    "scop3P_results = fetch_protein_modifications(TARGET_PROTEIN_ID)\n",
    "\n",
    "protein_name = scop3P_results['proteinName']\n",
    "entry_name = scop3P_results['entryName']\n",
    "accession = scop3P_results['accession']\n",
    "url = scop3P_results['url']\n",
    "modifications = scop3P_results['modifications']\n",
    "\n",
    "print(\"\"\"\n",
    "--------------------------------------------------------------------------------\n",
    "Scop3P: A Comprehensive Resource of Human Phosphosites within Their Full Context\n",
    "--------------------------------------------------------------------------------\n",
    "\n",
    "{0}:{1}\n",
    "\n",
    "Phospho-sites found: {3} entries.\n",
    "\n",
    "Full entry available on SCOP3P website: {2}\n",
    "\"\"\".format(entry_name, protein_name, url, len(modifications)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "51c73c52-5955-45d9-9d9d-fa3fe6069bfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>name</th>\n",
       "      <th>evidence</th>\n",
       "      <th>position</th>\n",
       "      <th>source</th>\n",
       "      <th>reference</th>\n",
       "      <th>functionalScore</th>\n",
       "      <th>specificSinglyPhosphorylated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>THR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>None</td>\n",
       "      <td>291</td>\n",
       "      <td>PRIDE</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>None</td>\n",
       "      <td>606</td>\n",
       "      <td>PRIDE</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SER</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Combined</td>\n",
       "      <td>696</td>\n",
       "      <td>UP, PRIDE</td>\n",
       "      <td>PubMed:19369195</td>\n",
       "      <td>0.641038</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SER</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>None</td>\n",
       "      <td>699</td>\n",
       "      <td>PRIDE</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>806</td>\n",
       "      <td>UP</td>\n",
       "      <td>PubMed:14711813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>809</td>\n",
       "      <td>UP</td>\n",
       "      <td>PubMed:14711813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>900</td>\n",
       "      <td>UP</td>\n",
       "      <td>PubMed:14711813,PubMed:16928683</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>905</td>\n",
       "      <td>UP</td>\n",
       "      <td>PubMed:20117004,PubMed:14711813,PubMed:1692868...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TYR</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>981</td>\n",
       "      <td>UP</td>\n",
       "      <td>PubMed:14711813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  residue             name      evidence  position     source  \\\n",
       "0     THR  phosphorylation          None       291      PRIDE   \n",
       "1     TYR  phosphorylation          None       606      PRIDE   \n",
       "2     SER  phosphorylation      Combined       696  UP, PRIDE   \n",
       "3     SER  phosphorylation          None       699      PRIDE   \n",
       "4     TYR  phosphorylation  Experimental       806         UP   \n",
       "5     TYR  phosphorylation  Experimental       809         UP   \n",
       "6     TYR  phosphorylation  Experimental       900         UP   \n",
       "7     TYR  phosphorylation  Experimental       905         UP   \n",
       "8     TYR  phosphorylation  Experimental       981         UP   \n",
       "\n",
       "                                           reference  functionalScore  \\\n",
       "0                                               None              NaN   \n",
       "1                                               None              NaN   \n",
       "2                                    PubMed:19369195         0.641038   \n",
       "3                                               None              NaN   \n",
       "4                                    PubMed:14711813              NaN   \n",
       "5                                    PubMed:14711813              NaN   \n",
       "6                    PubMed:14711813,PubMed:16928683              NaN   \n",
       "7  PubMed:20117004,PubMed:14711813,PubMed:1692868...              NaN   \n",
       "8                                    PubMed:14711813              NaN   \n",
       "\n",
       "   specificSinglyPhosphorylated  \n",
       "0                             1  \n",
       "1                             1  \n",
       "2                             1  \n",
       "3                             1  \n",
       "4                             0  \n",
       "5                             0  \n",
       "6                             0  \n",
       "7                             0  \n",
       "8                             0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_modification_table(modifications):\n",
    "    \"\"\"\n",
    "    Displays the protein modifications in a pandas DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    modifications (list): A list of dictionaries, each representing a protein modification.\n",
    "    \"\"\"\n",
    "    df = pd.DataFrame(modifications)\n",
    "    df = df[['residue', 'name', 'evidence', 'position', 'source', 'reference', 'functionalScore', 'specificSinglyPhosphorylated']]\n",
    "    \n",
    "    return df \n",
    "    \n",
    "modification_table = get_modification_table(modifications)\n",
    "display(modification_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98e42ffe-505f-4d29-a049-5bb56396b942",
   "metadata": {},
   "source": [
    "### Get Disease causing variants from UniProt using API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "353ffcec-1839-4083-8bca-16cc9370b723",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "\n",
    "ACC = \"P07949\"\n",
    "url = f\"https://www.ebi.ac.uk/proteins/api/variation/{ACC}\"\n",
    "headers = {\"Accept\": \"application/json\"}\n",
    "\n",
    "r = requests.get(url, headers=headers)\n",
    "r.raise_for_status()\n",
    "data = r.json()\n",
    "\n",
    "rows = []\n",
    "\n",
    "for feat in data.get(\"features\", []):\n",
    "    if feat.get(\"type\") != \"VARIANT\":\n",
    "        continue\n",
    "\n",
    "    for assoc in feat.get(\"association\", []):\n",
    "        if assoc.get(\"disease\") is not True:\n",
    "            continue\n",
    "\n",
    "        rows.append({\n",
    "            \"ACC_ID\": ACC,\n",
    "            \"position\": int(feat.get(\"begin\")),\n",
    "            \"WT\": feat.get(\"wildType\"),\n",
    "            \"MT\": feat.get(\"mutatedType\"),\n",
    "            \"consequence\": feat.get(\"consequenceType\"),\n",
    "            \"disease_name\": assoc.get(\"name\"),\n",
    "            \"disease_description\": assoc.get(\"description\"),\n",
    "            \"source_type\": feat.get(\"sourceType\")\n",
    "            # \"somatic_status\": feat.get(\"somaticStatus\"),\n",
    "            # \"genomic_location\": \";\".join(feat.get(\"genomicLocation\", [])),\n",
    "            # \"cytoband\": feat.get(\"cytogeneticBand\")\n",
    "        })\n",
    "\n",
    "mut_table = pd.DataFrame(rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a1113ac2-9418-4628-aa25-615bc8e277e0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mut_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07038b8d-ba1e-4f2e-9c60-74e1b353937e",
   "metadata": {},
   "source": [
    "## Rendering results\n",
    "\n",
    "For visualizing the 3D structure, we'll use py3Dmol library."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9702c23f",
   "metadata": {},
   "source": [
    "## Let's get AlphaFold model for the target protein\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6192377b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get alphaFold model for the protein\n",
    "import urllib.request\n",
    "AFurl=\"https://alphafold.ebi.ac.uk/files/AF-\"\n",
    "modelurl = f'{AFurl}{accession}{\"-F1-model_v6.pdb\"}'\n",
    "AFmodel = urllib.request.urlretrieve(modelurl,f'{accession}{\".pdb\"}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41b5d7c5",
   "metadata": {},
   "source": [
    "## Rendering results\n",
    "\n",
    "For visualizing the 3D structure, we'll use py3Dmol library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "157e10c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol\n",
    "\n",
    "def display_3D(modification_table, accession):\n",
    "    view = py3Dmol.view(width=700, height=500)\n",
    "    view.addModel(open(accession + '.pdb', 'r').read(), 'pdb')\n",
    "\n",
    "    view.setStyle({}, {'cartoon': {'color': 'silver'}})\n",
    "    view.addSurface(py3Dmol.VDW, {'opacity': 0.35, 'color': 'white'}, {})\n",
    "\n",
    "    # --- Color phosphosites \n",
    "    for _, row in modification_table.iterrows():\n",
    "        position = str(row['position'])\n",
    "\n",
    "        # Normalize residue label to avoid mismatches\n",
    "        residue = str(row['residue']).strip()  # removes trailing spaces etc.\n",
    "\n",
    "        if residue == 'TYR':\n",
    "            color = '#2CA02C'\n",
    "        elif residue == 'SER':\n",
    "            color = '#1F77B4'\n",
    "        elif residue == 'THR':\n",
    "            color = '#FF7F0E'\n",
    "        else:\n",
    "            color = '#7B241C'\n",
    "\n",
    "        sel = {'resi': position}  # add {'chain': row['chain']} if needed\n",
    "\n",
    "        view.addStyle(sel, {'stick': {'color': color}})\n",
    "        view.addStyle(sel, {'sphere': {'color': color, 'radius': 0.9}})\n",
    "\n",
    "    # --- Hover for ALL amino acids (all atoms) ---\n",
    "    view.setHoverable(\n",
    "        {}, True,\n",
    "        \"\"\"\n",
    "        function(atom, viewer, event, container) {\n",
    "            if(!atom.label) {\n",
    "                atom.label = viewer.addLabel(\n",
    "                    atom.resn + \" \" + atom.resi + (atom.chain ? (\" : \" + atom.chain) : \"\"),\n",
    "                    {position: atom, backgroundColor: 'mintcream', fontColor: 'black'}\n",
    "                );\n",
    "            }\n",
    "        }\n",
    "        \"\"\",\n",
    "        \"\"\"\n",
    "        function(atom, viewer) {\n",
    "            if(atom.label) {\n",
    "                viewer.removeLabel(atom.label);\n",
    "                delete atom.label;\n",
    "            }\n",
    "        }\n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "    view.zoomTo()\n",
    "    view.render()\n",
    "    return view\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "83b2f132",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_3D(modification_table,accession)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde07233",
   "metadata": {},
   "source": [
    "### PDB structure mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e15fc47d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_3D(modification_table, pdb_id, chain=None):\n",
    "    view = py3Dmol.view(query=f\"pdb:{pdb_id}\")\n",
    "\n",
    "    # Protein context\n",
    "    view.setStyle({}, {'cartoon': {'color': 'skyblue'}})\n",
    "\n",
    "    # Global surface (NO hover expected here)\n",
    "    view.addSurface(py3Dmol.VDW, {'opacity': 0.6, 'color': 'white'}, {})\n",
    "\n",
    "    # ---- Colored modified residues (ATOMS) ----\n",
    "    for _, row in modification_table.iterrows():\n",
    "        position = str(row['position'])\n",
    "        residue  = str(row['residue']).strip()\n",
    "\n",
    "        if residue == 'TYR':\n",
    "            color = '#2CA02C'\n",
    "        elif residue == 'SER':\n",
    "            color = '#1F77B4'\n",
    "        elif residue == 'THR':\n",
    "            color = '#FF7F0E'\n",
    "        else:\n",
    "            color = '#7B241C'\n",
    "\n",
    "        sel = {'resi': position}\n",
    "        if chain:\n",
    "            sel['chain'] = chain\n",
    "\n",
    "        # ATOMS → hover works\n",
    "        view.addStyle(sel, {'stick':  {'color': color}})\n",
    "        view.addStyle(sel, {'sphere': {'color': color, 'radius': 0.9}})\n",
    "\n",
    "    # ---- Hover for ALL amino acids ----\n",
    "    view.setHoverable(\n",
    "        {}, True,\n",
    "        \"\"\"\n",
    "        function(atom, viewer, event, container) {\n",
    "            if (!atom.label) {\n",
    "                atom.label = viewer.addLabel(\n",
    "                    atom.resn + \" \" + atom.resi + (atom.chain ? (\" : \" + atom.chain) : \"\"),\n",
    "                    {position: atom, backgroundColor: 'mintcream', fontColor: 'black'}\n",
    "                );\n",
    "            }\n",
    "        }\n",
    "        \"\"\",\n",
    "        \"\"\"\n",
    "        function(atom, viewer) {\n",
    "            if (atom.label) {\n",
    "                viewer.removeLabel(atom.label);\n",
    "                delete atom.label;\n",
    "            }\n",
    "        }\n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "    view.zoomTo()\n",
    "    view.render()\n",
    "    return view\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "695e2f23",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_3D(modification_table, pdb_id=\"6nja\", chain=\"A\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3673c56",
   "metadata": {},
   "source": [
    "## Mapping Biophysical properties on AF models\n",
    "\n",
    "For visualizing the 3D structure, we'll use py3Dmol library."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b2b826c",
   "metadata": {},
   "source": [
    "### Predict dynamic properties from Bio2Byte tools\n",
    "\n",
    "We will fetch the amino acid sequence from UniProtKB API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "054287d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_sequence_aminoacids(accession):\n",
    "    BASE_URL = f\"http://uniprot.org/uniprotkb/{accession}.fasta\"\n",
    "    url = f'{BASE_URL}?accession={accession}'\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        raw_fasta_sequence = response.content.decode(\"utf-8\")\n",
    "    else:\n",
    "        raw_fasta_sequence = \"\"\n",
    "    \n",
    "    lines = raw_fasta_sequence.split('\\n')\n",
    "    protein_id = str(lines[0])\n",
    "    amino_acids = \"\".join([str(l) for l in lines[1:]])\n",
    "    \n",
    "    return protein_id, amino_acids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "571a92ad",
   "metadata": {},
   "source": [
    "#### Predict Dynamic properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8c75dda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "_protein_id, sequence = fetch_sequence_aminoacids(TARGET_PROTEIN_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a1beaf87",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def predict_biophysical_features(accession, sequence):\n",
    "\n",
    "    with tempfile.NamedTemporaryFile(prefix=\"seq_\", suffix=\".fasta\", mode=\"w\") as fp:\n",
    "        fp.write(f\">{accession}\\n{sequence}\\n\")\n",
    "        fp.flush()\n",
    "        fp.seek(0)\n",
    "        \n",
    "        pred = SingleSeq(fp.name).predict(tools=[constants.TOOL_DYNAMINE, constants.TOOL_DISOMINE, constants.TOOL_EFOLDMINE]).get_all_predictions()\n",
    "    \n",
    "    return pred\n",
    "\n",
    "\n",
    "biophysical_features = predict_biophysical_features(TARGET_PROTEIN_ID, sequence)\n",
    "\n",
    "biophysical_features_target_protein = biophysical_features['proteins'][TARGET_PROTEIN_ID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "af0814fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# values above 1.0 membrane spanning regions, \n",
    "# Values above 0.8 indicate rigid conformations, \n",
    "# Values between 0.69-0.80 are 'context' dependent and capable of being either rigid or flexible.\n",
    "# values below 0.69 flexible regions. \n",
    "\n",
    "# phosphorylated = list(modifications_table['position'])\n",
    "biophysical_features_target_protein['BD_label'] = []\n",
    "biophysical_features_target_protein['DO_label'] = []\n",
    "biophysical_features_target_protein['EF_label'] = []\n",
    "# biophysical_features_target_protein['psites'] = []\n",
    "\n",
    "for index, residue in enumerate(biophysical_features_target_protein['seq']):\n",
    "    current_backbone = biophysical_features_target_protein['backbone'][index]\n",
    "    biophysical_features_target_protein['BD_label'].append(1 if current_backbone > 1 else 2 if current_backbone > 0.8 else 3 if current_backbone > 0.69 else 4)\n",
    "\n",
    "    current_disorder_propensity = biophysical_features_target_protein['disoMine'][index]\n",
    "    biophysical_features_target_protein['DO_label'].append(1 if current_disorder_propensity > 0.5 else 0)\n",
    "\n",
    "    current_early_folding = biophysical_features_target_protein['earlyFolding'][index]\n",
    "    biophysical_features_target_protein['EF_label'].append(1 if current_early_folding > 0.169 else 0)\n",
    "    \n",
    "#     biophysical_features_target_protein['psites'].append(1 if index + 1 in phosphorylated else 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dda3fd2",
   "metadata": {},
   "source": [
    "### Get the dynamic properties in a pandas DataFrame object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "dfe85850",
   "metadata": {},
   "outputs": [],
   "source": [
    "dynamic_properties=pd.DataFrame(biophysical_features_target_protein)\n",
    "dynamic_properties['seqpos']=range(1,len(dynamic_properties)+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df80d78",
   "metadata": {},
   "source": [
    "### Map onto AF models and visualize\n",
    "\n",
    "For visualizing the 3D structure, we'll use py3Dmol library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1e066d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "import colorsys\n",
    "\n",
    "\n",
    "def pseudocolor(minval, maxval,val):\n",
    "    \"\"\" Convert predicted values min.....max in range Green...Yellow..RED \n",
    "        The colors correspond to Red and Green in the HSV colorspace\n",
    "    \"\"\"\n",
    "    minval,maxval=float(minval),float(maxval)\n",
    "    h = (float(maxval-val) / (maxval-minval)) * 120\n",
    "    r, g, b = colorsys.hsv_to_rgb(h/360, 1., 1.)\n",
    "    rgb=map(lambda x: int(255 * x), (r, g, b))\n",
    "    rgb=tuple(rgb)\n",
    "    rgb='0x%02x%02x%02x' % rgb\n",
    "    return rgb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "53203ebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remap(df):\n",
    "    BDcolor,EFcolor,DOcolor={},{},{}\n",
    "    seqpos=0\n",
    "    min_BD,max_BD=min(df.backbone),max(df.backbone)\n",
    "    min_DO,max_DO=min(df.disoMine),max(df.disoMine)\n",
    "    min_EF,max_EF=min(df.earlyFolding),max(df.earlyFolding)\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        seqpos+=1\n",
    "        BDrescol=pseudocolor(min_BD,max_BD,float(row.backbone))\n",
    "        DOrescol=pseudocolor(min_EF,max_EF,float(row.disoMine))\n",
    "        EFrescol=pseudocolor(min_EF,max_EF,float(row.earlyFolding))\n",
    "        BDcolor[seqpos]=BDrescol\n",
    "        DOcolor[seqpos]=DOrescol\n",
    "        EFcolor[seqpos]=EFrescol\n",
    "        \n",
    "    return BDcolor,EFcolor,DOcolor\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "650fb2a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_3D(dynamic_properties):\n",
    "    BDcolor, EFcolor, DOcolor = remap(dynamic_properties)\n",
    "    modpos = modification_table.position.tolist()\n",
    "\n",
    "    view = py3Dmol.view(viewergrid=(2,2))\n",
    "    view.addModel(open((accession + '.pdb'), 'r').read(), 'pdb')\n",
    "\n",
    "    # IMPORTANT: setStyle(selection, style)\n",
    "    view.setStyle({}, {'cartoon': {'colorscheme': {'prop':'b','gradient':'rwb','min':0.0,'max':100.0}}}, viewer=(0,0))\n",
    "    view.setStyle({}, {'cartoon': {'colorscheme': {'prop':'resi','map':BDcolor}}}, viewer=(0,1))\n",
    "    view.setStyle({}, {'cartoon': {'colorscheme': {'prop':'resi','map':DOcolor}}}, viewer=(1,0))\n",
    "    view.setStyle({}, {'cartoon': {'colorscheme': {'prop':'resi','map':EFcolor}}}, viewer=(1,1))\n",
    "\n",
    "    # Surface highlight + pickable overlay on mod residues\n",
    "    for mod in modpos:\n",
    "        m = str(mod)\n",
    "        sel = {'resi': m}\n",
    "\n",
    "        view.addSurface(py3Dmol.VDW, {'opacity': 1.0}, sel, viewer=(0,0))\n",
    "        view.addSurface(py3Dmol.VDW, {'opacity': 1.0, 'color': BDcolor[mod]}, sel, viewer=(0,1))\n",
    "        view.addSurface(py3Dmol.VDW, {'opacity': 1.0, 'color': DOcolor[mod]}, sel, viewer=(1,0))\n",
    "        view.addSurface(py3Dmol.VDW, {'opacity': 1.0, 'color': EFcolor[mod]}, sel, viewer=(1,1))\n",
    "\n",
    "        # MAKE IT PICKABLE: opacity must be > 0\n",
    "        for panel in [(0,0), (0,1), (1,0), (1,1)]:\n",
    "            view.addStyle(sel, {'sphere': {'radius': 0.8, 'opacity': 0.15}}, viewer=panel)\n",
    "            # optional: stick helps pickability even more\n",
    "            # view.addStyle(sel, {'stick': {'opacity': 0.15}}, viewer=panel)\n",
    "\n",
    "    # Background + hover everywhere (per panel)\n",
    "    for panel in [(0,0), (0,1), (1,0), (1,1)]:\n",
    "        view.setBackgroundColor('white', viewer=panel)\n",
    "\n",
    "        view.setHoverable(\n",
    "            {},  # hover everywhere\n",
    "            True,\n",
    "            \"\"\"\n",
    "            function(atom, viewer, event, container) {\n",
    "                if (!atom.label) {\n",
    "                    atom.label = viewer.addLabel(\n",
    "                        atom.resn + \" \" + atom.resi + (atom.chain ? (\" : \" + atom.chain) : \"\"),\n",
    "                        {position: atom, backgroundColor: 'mintcream', fontColor:'black'}\n",
    "                    );\n",
    "                }\n",
    "            }\n",
    "            \"\"\",\n",
    "            \"\"\"\n",
    "            function(atom, viewer) {\n",
    "                if (atom.label) {\n",
    "                    viewer.removeLabel(atom.label);\n",
    "                    delete atom.label;\n",
    "                }\n",
    "            }\n",
    "            \"\"\",\n",
    "            viewer=panel\n",
    "        )\n",
    "\n",
    "    view.zoomTo()\n",
    "    view.render()\n",
    "    return view\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fe3987-b6d1-4bb0-8aae-7c1e088d3e8e",
   "metadata": {},
   "source": [
    "## Visualize the 3D structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "12d487e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_3D(dynamic_properties)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15c9ab07-dc90-43b1-80f6-8747df1c2194",
   "metadata": {},
   "source": [
    "### Let's convert the 3D protein structure into a Residue Interaction Network (2.5D map)\n",
    "\n",
    "Every amino acid is a node and an edge is drawn betweentwo amino acids if they are close (<=8 A ngstrom) in 3D space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "67836388-e788-45d1-8581-88491e542560",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import IFrame, display\n",
    "\n",
    "def nx_rin_to_pyvis_default(\n",
    "    G,\n",
    "    ptm_positions=None,\n",
    "    mutation_positions=None,\n",
    "    out_html=\"rin_pyvis.html\",\n",
    "    height=\"600px\",\n",
    "    width=\"100%\",\n",
    "    default_color=\"#B0B0B0\",   # light grey\n",
    "    ptm_color=\"#1f77b4\",       # blue\n",
    "    mut_color=\"#d62728\",       # red\n",
    "    both_color=\"#9467bd\",      # purple\n",
    "    node_size=15,\n",
    "    ptm_size=22,\n",
    "    mut_size=22,\n",
    "    both_size=26,\n",
    "    select_menu=True,\n",
    "    filter_menu=True\n",
    "):\n",
    "    ptm_set = set(int(x) for x in (ptm_positions or []))\n",
    "    mut_set = set(int(x) for x in (mutation_positions or []))\n",
    "\n",
    "    net = Network(\n",
    "        height=height,\n",
    "        width=width,\n",
    "        directed=False,\n",
    "        notebook=True,\n",
    "        cdn_resources=\"in_line\",\n",
    "        select_menu=select_menu,\n",
    "        filter_menu=filter_menu\n",
    "    )\n",
    "\n",
    "    net.set_options(\"\"\"\n",
    "    {\n",
    "      \"groups\": {\n",
    "        \"PTM\": {\n",
    "          \"color\": {\n",
    "            \"background\": \"#d62728\",\n",
    "            \"border\": \"#d62728\",\n",
    "            \"highlight\": { \"background\": \"#d62728\", \"border\": \"#d62728\" },\n",
    "            \"hover\":     { \"background\": \"#d62728\", \"border\": \"#d62728\" }\n",
    "          }\n",
    "        },\n",
    "        \"Mutation\": {\n",
    "          \"color\": {\n",
    "            \"background\": \"#1f77b4\",\n",
    "            \"border\": \"#1f77b4\",\n",
    "            \"highlight\": { \"background\": \"#1f77b4\", \"border\": \"#1f77b4\" },\n",
    "            \"hover\":     { \"background\": \"#1f77b4\", \"border\": \"#1f77b4\" }\n",
    "          }\n",
    "        },\n",
    "        \"PTM+Mutation\": {\n",
    "          \"color\": {\n",
    "            \"background\": \"#2ca02c\",\n",
    "            \"border\": \"#2ca02c\",\n",
    "            \"highlight\": { \"background\": \"#2ca02c\", \"border\": \"#2ca02c\" },\n",
    "            \"hover\":     { \"background\": \"#2ca02c\", \"border\": \"#2ca02c\" }\n",
    "          }\n",
    "        },\n",
    "        \"Other\": {\n",
    "          \"color\": {\n",
    "            \"background\": \"#9FA8B0\",\n",
    "            \"border\": \"#9FA8B0\",\n",
    "            \"highlight\": { \"background\": \"#9FA8B0\", \"border\": \"#9FA8B0\" },\n",
    "            \"hover\":     { \"background\": \"#9FA8B0\", \"border\": \"#9FA8B0\" }\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \"\"\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # ---- Nodes ----\n",
    "    for (ch, resi), attrs in G.nodes(data=True):\n",
    "        resi = int(resi)\n",
    "        resn = attrs.get(\"ResName\", \"\")\n",
    "\n",
    "        is_ptm = resi in ptm_set\n",
    "        is_mut = resi in mut_set\n",
    "\n",
    "        if is_ptm and is_mut:\n",
    "            bg = both_color\n",
    "            size = both_size\n",
    "            group = \"PTM+Mutation\"\n",
    "        elif is_mut:\n",
    "            bg = mut_color\n",
    "            size = mut_size\n",
    "            group = \"Mutation\"\n",
    "        elif is_ptm:\n",
    "            bg = ptm_color\n",
    "            size = ptm_size\n",
    "            group = \"PTM\"\n",
    "        else:\n",
    "            bg = default_color\n",
    "            size = node_size\n",
    "            group = \"Other\"\n",
    "\n",
    "        node_id = f\"{ch}:{resi}\"\n",
    "\n",
    "        net.add_node(\n",
    "            node_id,\n",
    "            label=f\"{resn} {resi}\" if resn else str(resi),\n",
    "            title=f\"Chain: {ch}<br>Residue: {resn}<br>Position: {resi}<br>Group: {group}\",\n",
    "            color={\n",
    "                \"background\": bg,\n",
    "                \"border\": \"#333333\",\n",
    "                \"highlight\": {\"background\": bg, \"border\": \"#000000\"},\n",
    "                \"hover\": {\"background\": bg, \"border\": \"#000000\"}\n",
    "            },\n",
    "            size=size,\n",
    "            group=group,\n",
    "            font={\"size\": 12}\n",
    "        )\n",
    "\n",
    "    # ---- Edges ----\n",
    "    for (ch1, r1), (ch2, r2), eattrs in G.edges(data=True):\n",
    "        a = f\"{ch1}:{int(r1)}\"\n",
    "        b = f\"{ch2}:{int(r2)}\"\n",
    "        dist = eattrs.get(\"distance\", None)\n",
    "\n",
    "        net.add_edge(\n",
    "            a, b,\n",
    "            color=\"#A9A9A9\",\n",
    "            title=f\"distance: {dist:.2f} Å\" if dist is not None else \"\"\n",
    "        )\n",
    "\n",
    "    html = net.generate_html()\n",
    "    with open(out_html, \"w\", encoding=\"utf-8\") as f:\n",
    "        f.write(html)\n",
    "\n",
    "    display(IFrame(out_html, width=900, height=650))\n",
    "    return out_html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccac0b3d-f6cb-4685-a498-0aa1101f27b4",
   "metadata": {},
   "source": [
    "### Visualize the RIN (Residue intercation Network) \n",
    "\n",
    "1. Phospho sites are colored red\n",
    "2. Mutations are colored blue\n",
    "3. Mutations + Phosphorylation colored Green"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "cc07fff1-c427-4bac-b4c2-2e04410e1e20",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptm_positions = list(set(modification_table[\"position\"].dropna().astype(int).tolist()))\n",
    "\n",
    "# Example mutation table with column 'position'\n",
    "mutation_positions = list(set(mut_table[\"position\"].dropna().astype(int).tolist()))\n",
    "\n",
    "nx_rin_to_pyvis_default(\n",
    "    G,\n",
    "    ptm_positions=ptm_positions,\n",
    "    mutation_positions=mutation_positions,\n",
    "    out_html=\"rin_pyvis.html\",\n",
    "    select_menu=True,\n",
    "    filter_menu=False\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
