{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a5dd7d4-467d-4aba-9355-04b482638ea7",
   "metadata": {},
   "source": [
    "# Scop3P\n",
    "\n",
    "A comprehensive database of human phosphosites within their full context. Scop3P integrates sequences (UniProtKB/Swiss-Prot), structures (PDB), and uniformly reprocessed phosphoproteomics data (PRIDE) to annotate all known human phosphosites. \n",
    "\n",
    "Scop3P, available at https://iomics.ugent.be/scop3p, presents a unique resource for visualization and analysis of phosphosites and for understanding of phosphosite structure–function relationships.\n",
    "\n",
    "Please cite: https://doi.org/10.1021/acs.jproteome.0c00306"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb1e9ff-c789-4c96-b71d-5da2af53b437",
   "metadata": {},
   "source": [
    "## Install Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ba367321-b7ac-44b9-98c0-c800fd1ae008",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install bokeh pandas matplotlib b2btools==3.0.7b2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62a30299-1084-4700-a8ed-64f77b68c0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "import tempfile\n",
    "import requests\n",
    "import pandas as pd \n",
    "import py3Dmol\n",
    "from b2bTools import SingleSeq, constants\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2dc9e1f-97d2-45a8-a21a-b7258664d265",
   "metadata": {},
   "source": [
    "## Target protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bd5e0bc6-c9ee-485f-9a7c-3adc53f7cd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_PROTEIN_ID = \"P07949\" # Write here the Protein ID of your protein of interest\n",
    "PDB_ID = \"2IVS\" # Write here the PDB ID of your protein of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e622d9b-1922-4851-ab6b-41ae0d62c395",
   "metadata": {},
   "source": [
    "## API Request\n",
    "\n",
    "This function makes a GET request to SCOP3P API endpoint for a given protein accession ID and returns the protein sequence in string format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87d604d4-ba1a-4acb-94b7-0481abe40fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_sequence_aminoacids(accession):\n",
    "    BASE_URL = f\"http://uniprot.org/uniprotkb/{accession}.fasta\"\n",
    "    url = f'{BASE_URL}?accession={accession}'\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        raw_fasta_sequence = response.content.decode(\"utf-8\")\n",
    "    else:\n",
    "        raw_fasta_sequence = \"\"\n",
    "    \n",
    "    lines = raw_fasta_sequence.split('\\n')\n",
    "    protein_id = str(lines[0])\n",
    "    amino_acids = \"\".join([str(l) for l in lines[1:]])\n",
    "    \n",
    "    return protein_id, amino_acids\n",
    "\n",
    "def fetch_protein_modifications(accession):\n",
    "    \"\"\"\n",
    "    Fetches protein modifications for a given UniProt ID.\n",
    "\n",
    "    Parameters:\n",
    "    accession (str): UniProt ID of the protein.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary containing protein modifications.\n",
    "    \"\"\"\n",
    "    BASE_URL = \"https://iomics.ugent.be/scop3p/api/modifications\"\n",
    "    url = f'{BASE_URL}?accession={accession}'\n",
    "    headers = {'accept': 'application/json'}\n",
    "    response = requests.get(url, headers=headers)\n",
    "    if response.status_code == 200:\n",
    "        return response.json()\n",
    "    else:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09ac2da-0a01-4b83-97ba-96c9fa9fb598",
   "metadata": {},
   "source": [
    "## Data parsing\n",
    "\n",
    "For parsing the JSON as a table, we'll use pandas library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fdd2c83a-bbdf-4ccd-993f-92078980fd9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Scop3P: A Comprehensive Resource of Human Phosphosites within Their Full Context\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "RET_HUMAN:Proto-oncogene tyrosine-protein kinase receptor Ret (EC 2.7.10.1) (Cadherin family member 12) (Proto-oncogene c-Ret) [Cleaved into: Soluble RET kinase fragment; Extracellular cell-membrane anchored RET cadherin 120 kDa fragment]\n",
      "\n",
      "Phospho-sites found: 9 entries.\n",
      "\n",
      "Full entry available on SCOP3P website: https://iomics.ugent.be/scop3p/index?protein=P07949\n",
      "\n"
     ]
    }
   ],
   "source": [
    "scop3P_results = fetch_protein_modifications(TARGET_PROTEIN_ID)\n",
    "_protein_id, sequence = fetch_sequence_aminoacids(TARGET_PROTEIN_ID)\n",
    "\n",
    "protein_name = scop3P_results['proteinName']\n",
    "entry_name = scop3P_results['entryName']\n",
    "accession = scop3P_results['accession']\n",
    "url = scop3P_results['url']\n",
    "modifications = scop3P_results['modifications']\n",
    "\n",
    "print(\"\"\"\n",
    "--------------------------------------------------------------------------------\n",
    "Scop3P: A Comprehensive Resource of Human Phosphosites within Their Full Context\n",
    "--------------------------------------------------------------------------------\n",
    "\n",
    "{0}:{1}\n",
    "\n",
    "Phospho-sites found: {3} entries.\n",
    "\n",
    "Full entry available on SCOP3P website: {2}\n",
    "\"\"\".format(entry_name, protein_name, url, len(modifications)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51c73c52-5955-45d9-9d9d-fa3fe6069bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_modifications_table(modifications):\n",
    "    \"\"\"\n",
    "    Displays the protein modifications in a pandas DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    modifications (list): A list of dictionaries, each representing a protein modification.\n",
    "    \"\"\"\n",
    "    df = pd.DataFrame(modifications)\n",
    "    df = df[['residue', 'name', 'evidence', 'position', 'source', 'reference', 'functionalScore', 'specificSinglyPhosphorylated']]\n",
    "    \n",
    "    return df \n",
    "    \n",
    "modifications_table = get_modifications_table(modifications)\n",
    "display(modifications_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac2ba7e8-8398-4840-89ba-5b58243710e1",
   "metadata": {},
   "source": [
    "## Predict biophysical features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6db8fd71-3716-45fe-a1fd-bbed40f7d8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "def predict_biophysical_features(accession, sequence):\n",
    "    with tempfile.NamedTemporaryFile(prefix=\"seq_\", suffix=\".fasta\", mode=\"w\") as fp:\n",
    "        fp.write(f\">{accession}\\n{sequence}\\n\")\n",
    "        fp.flush()\n",
    "        fp.seek(0)\n",
    "        \n",
    "        pred = SingleSeq(fp.name).predict(tools=[constants.TOOL_DYNAMINE, constants.TOOL_DISOMINE, constants.TOOL_EFOLDMINE]).get_all_predictions()\n",
    "    \n",
    "    return pred\n",
    "\n",
    "biophysical_features = predict_biophysical_features(TARGET_PROTEIN_ID, sequence)\n",
    "\n",
    "biophysical_features_target_protein = biophysical_features['proteins'][TARGET_PROTEIN_ID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04ef2539-f2f4-462c-81a3-3214f2ec1468",
   "metadata": {},
   "outputs": [],
   "source": [
    "phosphorylated = list(modifications_table['position'])\n",
    "biophysical_features_target_protein['BD_label'] = []\n",
    "biophysical_features_target_protein['DO_label'] = []\n",
    "biophysical_features_target_protein['EF_label'] = []\n",
    "biophysical_features_target_protein['psites'] = []\n",
    "\n",
    "for index, residue in enumerate(biophysical_features_target_protein['seq']):\n",
    "    current_backbone = biophysical_features_target_protein['backbone'][index]\n",
    "    biophysical_features_target_protein['BD_label'].append(1 if current_backbone > 1 else 2 if current_backbone > 0.8 else 3 if current_backbone > 0.69 else 4)\n",
    "\n",
    "    current_disorder_propensity = biophysical_features_target_protein['disoMine'][index]\n",
    "    biophysical_features_target_protein['DO_label'].append(1 if current_disorder_propensity > 0.5 else 0)\n",
    "\n",
    "    current_early_folding = biophysical_features_target_protein['earlyFolding'][index]\n",
    "    biophysical_features_target_protein['EF_label'].append(1 if current_early_folding > 0.169 else 0)\n",
    "    \n",
    "    biophysical_features_target_protein['psites'].append(1 if index + 1 in phosphorylated else 0)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33730cd5",
   "metadata": {},
   "source": [
    "### Get the dynamic properties in a pandas DataFrame object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "30b1451f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dynamic_properties=pd.DataFrame(biophysical_features_target_protein)\n",
    "dynamic_properties['seqpos']=range(1,len(dynamic_properties)+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fdfb089",
   "metadata": {},
   "source": [
    "## Plotting results\n",
    "\n",
    "For visualization of dynamic properties we will use Bokeh interactive visualization package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b2ea610",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_notebook()\n",
    "\n",
    "p = figure(width=1000, height=300,tools = \"pan,box_zoom,reset\",toolbar_location=\"below\",\n",
    "           toolbar_sticky=False)\n",
    "p.title.text = 'Biophysical properties'\n",
    "\n",
    "\n",
    "l1=p.line(dynamic_properties.seqpos, dynamic_properties.backbone, line_width=2, color='blue', alpha=0.8,\n",
    "        muted_color='blue', muted_alpha=0.2, legend_label='backbone_dynamics')\n",
    "\n",
    "l2=p.line(dynamic_properties.seqpos, dynamic_properties.disoMine, line_width=2, color='red', alpha=0.8,\n",
    "        muted_color='red', muted_alpha=0.2, legend_label='disorder')\n",
    "\n",
    "l3=p.line(dynamic_properties.seqpos, dynamic_properties.earlyFolding, line_width=2, color='grey', alpha=0.8,\n",
    "        muted_color='grey', muted_alpha=0.2, legend_label='earlyFolding')\n",
    "\n",
    "l4=p.scatter(modifications_table.position,[0.5]*len(modifications_table.position),\\\n",
    "             fill_alpha=0.6, color='grey',size=10,legend_label='P-sites')\n",
    "\n",
    "p.add_tools(HoverTool(tooltips=\"Seqpos:@x, value:@y\", renderers=[l1,l2,l3]))\n",
    "p.add_tools(HoverTool(tooltips=\"Seqpos:@x\", renderers=[l4]))\n",
    "# p.add_tools(BoxZoomTool(match_aspect=False))\n",
    "\n",
    "\n",
    "legend=p.legend[0]\n",
    "p.legend.click_policy=\"mute\"\n",
    "p.add_layout(legend, 'right')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17952635-13d3-4c2a-80c2-9f1387510e88",
   "metadata": {},
   "source": [
    "## visualize the plots and interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d2c61430-d6d0-47d5-bd82-e54f352581d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.models import Div\n",
    "from bokeh.layouts import column\n",
    "\n",
    "notes = Div(\n",
    "    width=1000,\n",
    "    style={\"padding\": \"10px\"},\n",
    "    text=\"\"\"\n",
    "    <div style=\"font-family: Arial; font-size: 12px; line-height: 1.4;\">\n",
    "\n",
    "      <b>How to read the tracks</b><br><br>\n",
    "\n",
    "      <b style=\"color: blue;\">Backbone dynamics</b>:\n",
    "      <span style=\"color: blue;\">\n",
    "        &gt;1.0 membrane-spanning,\n",
    "        0.8–1.0 rigid,\n",
    "        0.69–0.80 context-dependent,\n",
    "        &lt;0.69 flexible\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: red;\">Disorder (DisoMine)</b>:\n",
    "      <span style=\"color: red;\">\n",
    "        values &gt;0.50 indicate disordered regions\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: grey;\">Early folding</b>:\n",
    "      <span style=\"color: grey;\">\n",
    "        values &gt;0.169 suggest early-folding propensity\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: grey;\">P-sites</b>:\n",
    "      <span style=\"color: grey;\">\n",
    "        phosphorylation positions (grey dots)\n",
    "      </span>\n",
    "\n",
    "    </div>\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "show(column(p, notes))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04ef888f",
   "metadata": {},
   "source": [
    "## Mutation analysis on Target protein\n",
    "\n",
    "We will try to mutate some amino acids in one or multiple positions and \n",
    "predict the Biophysical properties for the mutated sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "edb9e847",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global store (used later by inference tables)\n",
    "MUTATIONS = []\n",
    "\n",
    "def mutateSeq_and_store(sequence, store_global=True):\n",
    "    \"\"\"\n",
    "    Ask once for mutation positions + mutant AAs.\n",
    "    Returns:\n",
    "      mutated_sequence (str)\n",
    "      mutations (list of tuples): [(pos_1based, mutAA), ...]\n",
    "    Side effect (optional):\n",
    "      sets global MUTATIONS so later cells can reuse it without asking again.\n",
    "    \"\"\"\n",
    "\n",
    "    pos_raw = input(\"Please enter the positions to mutate (comma-separated, 1-based): \").strip()\n",
    "    aa_raw  = input(\"Please enter the aminoacid(s) to mutate TO (comma-separated): \").strip()\n",
    "\n",
    "    poslis = [p.strip() for p in pos_raw.split(\",\") if p.strip()]\n",
    "    aalis  = [a.strip().upper() for a in aa_raw.split(\",\") if a.strip()]\n",
    "\n",
    "    if len(poslis) != len(aalis):\n",
    "        raise ValueError(\"Number of positions and amino acids must match.\")\n",
    "\n",
    "    standardAA = set(list(\"ARNDCEQGHILKMFPSTWYV\"))\n",
    "    bad = [a for a in aalis if a not in standardAA]\n",
    "    if bad:\n",
    "        raise ValueError(f\"Some amino acid(s) not recognized: {bad}. Use standard 1-letter codes.\")\n",
    "\n",
    "    positions = [int(p) for p in poslis]\n",
    "    L = len(sequence)\n",
    "\n",
    "    # Validate positions\n",
    "    out_of_range = [p for p in positions if p < 1 or p > L]\n",
    "    if out_of_range:\n",
    "        raise ValueError(f\"Position(s) out of range (1..{L}): {out_of_range}\")\n",
    "\n",
    "    # Apply mutations (supports multiple)\n",
    "    mut_seq_list = list(sequence)\n",
    "    mutations = []\n",
    "\n",
    "    # If user repeats a position, last one wins\n",
    "    for pos, aa in zip(positions, aalis):\n",
    "        mut_seq_list[pos - 1] = aa\n",
    "        mutations.append((pos, aa))\n",
    "\n",
    "    mutated_sequence = \"\".join(mut_seq_list)\n",
    "\n",
    "    if store_global:\n",
    "        global MUTATIONS\n",
    "        MUTATIONS = mutations\n",
    "\n",
    "    # Quick summary\n",
    "    summary = \", \".join([f\"{sequence[p-1]}{p}{aa}\" for p, aa in mutations])\n",
    "    print(\"Applied mutations:\", summary)\n",
    "\n",
    "    return mutated_sequence, mutations\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f9f90ea",
   "metadata": {},
   "source": [
    "## Mutate here\n",
    "Now let's try to enter where we want to mutate and replace the amino acids at the position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0d1cd5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutated_sequence, MUTATIONS = mutateSeq_and_store(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eef1861f",
   "metadata": {},
   "source": [
    "### Predict Biophysical properties for the mutated sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab18ae04",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "\n",
    "biophysical_features_mut = predict_biophysical_features(TARGET_PROTEIN_ID, mutated_sequence)\n",
    "\n",
    "biophysical_features_mut_protein = biophysical_features_mut['proteins'][TARGET_PROTEIN_ID]\n",
    "\n",
    "biophysical_features_mut_protein['BD_label'] = []\n",
    "biophysical_features_mut_protein['DO_label'] = []\n",
    "biophysical_features_mut_protein['EF_label'] = []\n",
    "\n",
    "for index, residue in enumerate(biophysical_features_mut_protein['seq']):\n",
    "    mut_backbone = biophysical_features_mut_protein['backbone'][index]\n",
    "    biophysical_features_mut_protein['BD_label'].append(1 if mut_backbone > 1 else 2 if mut_backbone > 0.8 else 3 if mut_backbone > 0.69 else 4)\n",
    "\n",
    "    mut_disorder_propensity = biophysical_features_mut_protein['disoMine'][index]\n",
    "    biophysical_features_mut_protein['DO_label'].append(1 if mut_disorder_propensity > 0.5 else 0)\n",
    "\n",
    "    mut_early_folding = biophysical_features_mut_protein['earlyFolding'][index]\n",
    "    biophysical_features_mut_protein['EF_label'].append(1 if mut_early_folding > 0.169 else 0)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d01fb9a",
   "metadata": {},
   "source": [
    "### Covert biophysical properties into a pandas DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3bfefcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dynamic_properties_mut=pd.DataFrame(biophysical_features_mut_protein)\n",
    "dynamic_properties_mut['seqpos']=range(1,len(dynamic_properties_mut)+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c507682",
   "metadata": {},
   "source": [
    "### Plot results of mutated sequence\n",
    "Let's plot the Biophysical properties for the Original and Mutated sequences\n",
    "\n",
    "For visualization of dynamic properties we will use Bokeh interactive visualization package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eee0c15d",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_notebook()\n",
    "\n",
    "p_mut = figure(width=1000, height=300,tools = \"pan,box_zoom,reset\",toolbar_location=\"below\",\n",
    "           toolbar_sticky=False)\n",
    "p_mut.title.text = 'Biophysical properties'\n",
    "\n",
    "\n",
    "b1=p_mut.line(dynamic_properties.seqpos, dynamic_properties.backbone, line_width=2, color='skyblue', alpha=0.8,\n",
    "        muted_color='skyblue', muted_alpha=0.2, legend_label='backbone_dynamics')\n",
    "\n",
    "b2=p_mut.line(dynamic_properties_mut.seqpos, dynamic_properties_mut.backbone, line_width=2, color='blue', alpha=0.8,\n",
    "        muted_color='blue', muted_alpha=0.2, legend_label='backbone_mut')\n",
    "\n",
    "\n",
    "d1=p_mut.line(dynamic_properties.seqpos, dynamic_properties.disoMine, line_width=2, color='salmon', alpha=0.8,\n",
    "        muted_color='salmon', muted_alpha=0.2, legend_label='disorder')\n",
    "d2=p_mut.line(dynamic_properties_mut.seqpos, dynamic_properties_mut.disoMine, line_width=2, color='red', alpha=0.8,\n",
    "        muted_color='red', muted_alpha=0.2, legend_label='disorder_mut')\n",
    "\n",
    "e1=p_mut.line(dynamic_properties.seqpos, dynamic_properties.earlyFolding, line_width=2, color='grey', alpha=0.8,\n",
    "        muted_color='grey', muted_alpha=0.2, legend_label='earlyFolding')\n",
    "e2=p_mut.line(dynamic_properties_mut.seqpos, dynamic_properties_mut.earlyFolding, line_width=2, color='black', alpha=0.8,\n",
    "        muted_color='black', muted_alpha=0.2, legend_label='earlyFolding_mut')\n",
    "\n",
    "l4=p_mut.scatter(modifications_table.position,[0.5]*len(modifications_table.position),\\\n",
    "             fill_alpha=0.6, color='grey',size=10,legend_label='P-sites')\n",
    "\n",
    "p_mut.add_tools(HoverTool(tooltips=\"Seqpos:@x, value:@y\", renderers=[b1,b2,d1,d2,e1,e2]))\n",
    "p_mut.add_tools(HoverTool(tooltips=\"Seqpos:@x\", renderers=[l4]))\n",
    "\n",
    "\n",
    "legend=p_mut.legend[0]\n",
    "p_mut.legend.click_policy=\"mute\"\n",
    "p_mut.add_layout(legend, 'right')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9b78e4a1-ec3c-4801-8c15-2baefda72e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.models import Div\n",
    "from bokeh.layouts import column\n",
    "\n",
    "notes = Div(\n",
    "    width=1000,\n",
    "    style={\"padding\": \"10px\"},\n",
    "    text=\"\"\"\n",
    "    <div style=\"font-family: Arial; font-size: 12px; line-height: 1.4;\">\n",
    "\n",
    "      <b>How to read the tracks</b><br><br>\n",
    "\n",
    "      <b style=\"color: blue;\">Backbone dynamics</b>:\n",
    "      <span style=\"color: blue;\">\n",
    "        &gt;1.0 membrane-spanning,\n",
    "        0.8–1.0 rigid,\n",
    "        0.69–0.80 context-dependent,\n",
    "        &lt;0.69 flexible\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: red;\">Disorder (DisoMine)</b>:\n",
    "      <span style=\"color: red;\">\n",
    "        values &gt;0.50 indicate disordered regions\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: grey;\">Early folding</b>:\n",
    "      <span style=\"color: grey;\">\n",
    "        values &gt;0.169 suggest early-folding propensity\n",
    "      </span>\n",
    "      <br><br>\n",
    "\n",
    "      <b style=\"color: grey;\">P-sites</b>:\n",
    "      <span style=\"color: grey;\">\n",
    "        phosphorylation positions (grey dots)\n",
    "      </span>\n",
    "\n",
    "    </div>\n",
    "    \"\"\"\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "27cfd950",
   "metadata": {},
   "outputs": [],
   "source": [
    "show(column(p_mut,notes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd013272-379c-4491-a0fb-7914d1d75403",
   "metadata": {},
   "source": [
    "### Save the plot as svg, png, or jpg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9285696",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.output_backend = \"svg\"\n",
    "export_svgs(plot, filename=\"plot.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4eabd0-e8eb-42c6-ab7f-e2d42211755c",
   "metadata": {},
   "source": [
    "### Inference (What does the mutataion predictions tells us about the biophysical properties?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2c9c4d70-eae6-4d53-ac91-e7556fb2d119",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "\n",
    "# --- Label functions based on your cutoffs ---\n",
    "\n",
    "def label_backbone(x: float) -> str:\n",
    "    if x > 1.0:\n",
    "        return \"membrane-spanning\"\n",
    "    if x > 0.8:\n",
    "        return \"rigid\"\n",
    "    if x > 0.69:\n",
    "        return \"context-dependent\"\n",
    "    return \"flexible\"\n",
    "\n",
    "def label_disorder(x: float) -> str:\n",
    "    return \"disordered\" if x > 0.50 else \"ordered\"\n",
    "\n",
    "def label_earlyfold(x: float) -> str:\n",
    "    return \"early-folding\" if x > 0.169 else \"low\"\n",
    "\n",
    "LABEL_FUNCS = {\n",
    "    \"backbone\": (label_backbone, \"Backbone\"),\n",
    "    \"disoMine\": (label_disorder, \"Disorder\"),\n",
    "    \"earlyFolding\": (label_earlyfold, \"EarlyFolding\"),\n",
    "}\n",
    "\n",
    "def mutation_effect_table_with_label_shift(\n",
    "    wt_df: pd.DataFrame,\n",
    "    mut_df: pd.DataFrame,\n",
    "    feature: str,\n",
    "    window: int = 5,\n",
    "    mutations=None,   # uses MUTATIONS by default\n",
    "):\n",
    "    if mutations is None:\n",
    "        mutations = MUTATIONS\n",
    "\n",
    "    if feature not in LABEL_FUNCS:\n",
    "        raise ValueError(f\"feature must be one of {list(LABEL_FUNCS.keys())}\")\n",
    "\n",
    "    label_fn, pretty = LABEL_FUNCS[feature]\n",
    "\n",
    "    wt = wt_df.set_index(\"seqpos\", drop=False)\n",
    "    mu = mut_df.set_index(\"seqpos\", drop=False)\n",
    "    max_pos = int(wt[\"seqpos\"].max())\n",
    "\n",
    "    rows = []\n",
    "\n",
    "    for pos, aa_to in mutations:\n",
    "        if pos not in wt.index or pos not in mu.index:\n",
    "            rows.append({\n",
    "                \"pos\": pos, \"WT_AA\": \"\", \"Mut_AA\": aa_to,\n",
    "                \"mutation\": f\"?{pos}{aa_to}\", \"window\": \"\",\n",
    "                \"WT_label@pos\": \"\", \"Mut_label@pos\": \"\",\n",
    "                \"WT_label_mean\": \"\", \"Mut_label_mean\": \"\",\n",
    "                \"Δ@pos\": None, \"Δ_mean\": None,\n",
    "                \"inference\": \"Position out of range\", \"note\": \"\"\n",
    "            })\n",
    "            continue\n",
    "\n",
    "        wt_aa = str(wt.loc[pos, \"seq\"])\n",
    "        lo = max(1, pos - window)\n",
    "        hi = min(max_pos, pos + window)\n",
    "\n",
    "        wt_center = float(wt.loc[pos, feature])\n",
    "        mu_center = float(mu.loc[pos, feature])\n",
    "        d_center = mu_center - wt_center\n",
    "\n",
    "        wt_mean = float(wt.loc[lo:hi, feature].astype(float).mean())\n",
    "        mu_mean = float(mu.loc[lo:hi, feature].astype(float).mean())\n",
    "        d_mean = mu_mean - wt_mean\n",
    "\n",
    "        wt_lab_pos = label_fn(wt_center)\n",
    "        mu_lab_pos = label_fn(mu_center)\n",
    "        wt_lab_mean = label_fn(wt_mean)\n",
    "        mu_lab_mean = label_fn(mu_mean)\n",
    "\n",
    "        # Inference focuses on label transitions\n",
    "        if wt_lab_pos != mu_lab_pos:\n",
    "            shift_pos = f\"{wt_lab_pos} → {mu_lab_pos}\"\n",
    "        else:\n",
    "            shift_pos = f\"{wt_lab_pos} (no class change)\"\n",
    "\n",
    "        if wt_lab_mean != mu_lab_mean:\n",
    "            shift_mean = f\"{wt_lab_mean} → {mu_lab_mean}\"\n",
    "        else:\n",
    "            shift_mean = f\"{wt_lab_mean} (no class change)\"\n",
    "\n",
    "        inference = (\n",
    "            f\"{pretty}: {shift_pos} at site (Δ {d_center:+.3f}); \"\n",
    "            f\"window mean: {shift_mean} (Δ {d_mean:+.3f})\"\n",
    "        )\n",
    "\n",
    "        note_parts = []\n",
    "        # sanity check: if your mutated_sequence is used, this should match\n",
    "        mut_aa_seq = str(mu.loc[pos, \"seq\"])\n",
    "        if mut_aa_seq.upper() != aa_to.upper():\n",
    "            note_parts.append(f\"Mut seq AA={mut_aa_seq} (expected {aa_to})\")\n",
    "\n",
    "        rows.append({\n",
    "            \"pos\": pos,\n",
    "            \"WT_AA\": wt_aa,\n",
    "            \"Mut_AA\": aa_to,\n",
    "            \"mutation\": f\"{wt_aa}{pos}{aa_to}\",\n",
    "            \"window\": f\"{lo}-{hi}\",\n",
    "\n",
    "            f\"{pretty}_WT@pos\": wt_center,\n",
    "            f\"{pretty}_Mut@pos\": mu_center,\n",
    "            \"WT_label@pos\": wt_lab_pos,\n",
    "            \"Mut_label@pos\": mu_lab_pos,\n",
    "            \"Δ@pos\": d_center,\n",
    "\n",
    "            f\"{pretty}_WT_mean(±{window})\": wt_mean,\n",
    "            f\"{pretty}_Mut_mean(±{window})\": mu_mean,\n",
    "            \"WT_label_mean\": wt_lab_mean,\n",
    "            \"Mut_label_mean\": mu_lab_mean,\n",
    "            \"Δ_mean\": d_mean,\n",
    "\n",
    "            \"inference\": inference,\n",
    "            \"note\": \"; \".join(note_parts)\n",
    "        })\n",
    "\n",
    "    df = pd.DataFrame(rows)\n",
    "\n",
    "    # Pretty rounding for numeric cols\n",
    "    num_cols = [c for c in df.columns if any(k in c for k in [\"_WT@\", \"_Mut@\", \"_WT_mean\", \"_Mut_mean\", \"Δ@pos\", \"Δ_mean\"])]\n",
    "    for c in num_cols:\n",
    "        df[c] = pd.to_numeric(df[c], errors=\"coerce\").round(3)\n",
    "\n",
    "    # display(df.style.set_caption(f\"{pretty} label-shift summary (±{window} AA)\"))\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3f04448e-80af-4194-91be-f1aaeb0c770c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Backbone label shifts\n",
    "mutation_effect_table_with_label_shift(dynamic_properties, dynamic_properties_mut, feature=\"backbone\", mutations=MUTATIONS, window=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "99afa541-51eb-499f-94bf-6d5c1fb87639",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Disorder label shifts\n",
    "mutation_effect_table_with_label_shift(dynamic_properties, dynamic_properties_mut, feature=\"disoMine\", mutations=MUTATIONS, window=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4889517f-74bc-49ca-b797-5ffe1e9b7347",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Disorder label shifts\n",
    "mutation_effect_table_with_label_shift(dynamic_properties, dynamic_properties_mut, feature=\"earlyFolding\", mutations=MUTATIONS, window=5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
